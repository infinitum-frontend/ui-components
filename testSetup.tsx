import { expect, beforeAll } from 'vitest'
import { render, RenderOptions } from '@testing-library/react'
// import '@testing-library/jest-dom/vitest' for vitest 1.x.x
import { ReactElement, JSXElementConstructor } from 'react'
import matchers from '@testing-library/jest-dom/matchers'

// extends Vitest's expect method with methods from react-testing-library
// https://github.com/testing-library/jest-dom#custom-matchers
expect.extend(matchers)

export const renderComponent = function (
  Component: ReactElement,
  options?: RenderOptions
): {
  el: HTMLElement
  container: HTMLElement
  rerender: (ui: ReactElement<any, string | JSXElementConstructor<any>>) => void
} {
  const { container, rerender } = render(Component, options)

  const el = container.firstChild as HTMLElement

  return { el, container, rerender }
}

beforeAll(() => {
  global.ResizeObserver = class ResizeObserver {
    observe(): any {}
    unobserve(): any {}
    disconnect(): any {}
  }

  // https://github.com/testing-library/react-testing-library/issues/95#issuecomment-392879866
  // https://github.com/vuejs/vue-test-utils/issues/319
  if (!HTMLElement.prototype.scrollTo) {
    HTMLElement.prototype.scrollTo = () => {}
  }
})

// runs a cleanup after each test case (e.g. clearing jsdom)
// afterEach(() => {
//   cleanup()
// })
