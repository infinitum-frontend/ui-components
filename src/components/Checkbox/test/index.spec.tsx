import { describe, it, vi } from 'vitest'
import { renderComponent } from '../../../../testSetup'
import { Checkbox } from '../index'
import userEvent from '@testing-library/user-event'

describe('Checkbox', () => {
  const user = userEvent.setup()
  it('should render', () => {
    const { el } = renderComponent(<Checkbox />)
    const input = el.querySelector('input')
    expect(el).toBeInTheDocument()
    expect(el).toHaveClass('inf-checkbox')
    expect(input).toHaveProperty('type', 'checkbox')
  })

  it('should match snapshop', () => {
    const { el } = renderComponent(<Checkbox />)
    expect(el).toMatchSnapshot()
  })

  it('should be unchecked by default', () => {
    renderComponent(<Checkbox />)
    const input = document.querySelector('input')
    expect(input).toHaveProperty('checked', false)
  })

  it('should support uncontrolled variant', async () => {
    const { el } = renderComponent(<Checkbox defaultChecked={true} />)
    const input = el.querySelector('input')
    expect(input).toHaveProperty('checked', true)
    await user.click(el)
    expect(input).toHaveProperty('checked', false)
    await user.click(el)
    expect(input).toHaveProperty('checked', true)
  })

  it('should support controlled variant', async () => {
    let checked
    let event
    const onChange = vi.fn((value, e) => {
      checked = value
      event = e
    })
    const { el } = renderComponent(
      <Checkbox checked={true} onChange={onChange} />
    )
    const input = el.querySelector('input')
    expect(input).toHaveProperty('checked', true)
    await user.click(el)
    expect(onChange).toBeCalledTimes(1)
    expect(onChange).toBeCalledWith(checked, event)
  })

  // TODO: обратиться в vitest за поддержкой. Или в svgr?
  // it('should show icon on checked', () => {
  //   const { el } = renderComponent<CheckboxProps>(Checkbox, {
  //     defaultChecked: true
  //   })
  //
  //   const icon = el.querySelector('svg')
  //   console.log(getComputedStyle(icon))
  //   expect(icon).toBeVisible()
  //   expect(icon).toHaveStyle('visibility: visible')
  // })

  it('should support indeterminate', () => {
    const { el } = renderComponent(<Checkbox indeterminate={true} />)
    const box = el.querySelector('.inf-checkbox__box')
    expect(box).toHaveClass('inf-checkbox__box--indeterminate')
  })

  it('should support disabled', async () => {
    const onChange = vi.fn()
    const { el } = renderComponent(
      <Checkbox disabled={true} onChange={onChange}>
        Label
      </Checkbox>
    )
    const box = el.querySelector('.inf-checkbox__box')
    const input = el.querySelector('input')
    expect(input).toBeDisabled()
    expect(box).toHaveClass('inf-checkbox__box--disabled')
    expect(el).toHaveClass('inf-checkbox--disabled')

    await user.click(el).catch(() => {})
    expect(onChange).toHaveBeenCalledTimes(0)
  })

  it('should support name and value', () => {
    const { el } = renderComponent(<Checkbox name={'Name'} value={'Value'} />)
    const input = el.querySelector('input')
    expect(input).toHaveAttribute('name', 'Name')
    expect(input).toHaveAttribute('value', 'Value')
  })

  it('should support className', () => {
    const { el } = renderComponent(<Checkbox className={'custom-class'} />)
    expect(el).toHaveClass('custom-class')
    expect(el).toHaveClass('inf-checkbox')
  })

  it('should render children', () => {
    const { el } = renderComponent(<Checkbox>Label</Checkbox>)
    expect(el).toContainHTML('Label')
  })

  it('should support basic inputProps', () => {
    const { el } = renderComponent(
      <Checkbox inputProps={{ 'aria-checked': 'false' }}>Label</Checkbox>
    )
    const input = el.querySelector('input')

    expect(input).toHaveAttribute('aria-checked', 'false')
  })
})
