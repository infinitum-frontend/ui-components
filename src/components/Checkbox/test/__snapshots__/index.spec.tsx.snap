// Vitest Snapshot v1

exports[`Checkbox > should match snapshop 1`] = `
<label
  class="inf-checkbox"
>
  <input
    aria-checked="false"
    type="checkbox"
    value=""
  />
  <span
    class="inf-checkbox__box"
  >
    <svg
      fill="none"
      height="16px"
      viewBox="0 0 16 16"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 7.70588L7.11111 11L11.5 4"
        stroke="currentColor"
        stroke-linejoin="round"
        stroke-width="1.5"
      />
    </svg>
  </span>
</label>
`;
